File: ruby/lib/task_master/output_handlers/countdown_report/terminal_handler.rb
module TaskMaster
  module OutputHandlers
    module CountdownReport
      class TerminalHandler
        def generate_output(data)
          AsciiTable.new(headers, DataProcessor.new(data).process).draw
        end

        private

        def headers
          ["Task ID", "Task Name", "Assignee", "Due Date", "Status", "Time Remaining (Days)"]
        end

        class DataProcessor
          def initialize(data)
            @data = data
          end

          def process
            @data.map do |task|
              [
                task.id.to_s,
                task.name,
                task.assignee,
                task.due_on.to_s,
                task.status,
                time_until_due_string(task)
              ]
            end
          end

          private

          def time_until_due_string(task)
            if task.number_of_days_until_due.abs == 1
              task.overdue? ? "1 day overdue" : "1 day"
            elsif task.complete?
              "-"
            else
              "#{task.number_of_days_until_due.abs} days#{' overdue' if task.overdue?}"
            end
          end
        end
      end
    end
  end
end

File: javascript/lib/outputHandlers/countdownReport/TerminalHandler.js
const path = require('path');
const AsciiTable = require(path.join(__dirname, '../../AsciiTable'));

class TerminalHandler {
	constructor() {
		this.AsciiTable = AsciiTable
		this.DataProcessor = class {
			constructor(data) {
				this.data = data
			}

			process() {
				return this.data.map(task => {
					return [
						String(task.id),
						task.name,
						task.assignee,
						this.#formatDateString(task.dueOn),
						task.status,
						this.#timeUntilDueString(task)
					]
				})
			}

			#timeUntilDueString(task) {
				if(Math.abs(task.numberOfDaysUntilDue()) == 1) {
					return task.isOverdue() ? "1 day overdue" : "1 day";
				} else if(task.isComplete()) {
					return "-";
				} else {
					const overdue = ""
					if(task.isOverdue) {
						const overdue = ' overdue'
					}
					return `${task.numberOfDaysUntilDue()} days${overdue}`;
				}
			}

			#formatDateString(date) {
				return String(date.getFullYear() + '-' + (date.getMonth() + 1) + '-' + String(date.getDate()).padStart(2, '0'));
			}
		}
	}

	generateOutput(data) {
		return new this.AsciiTable(this.#headers(), new this.DataProcessor(data).process()).draw()
	}

	#headers() {
		return ["Task ID", "Task Name", "Assignee", "Due Date", "Status", "Time Remaining (Days)"]
	}
}

module.exports = TerminalHandler;
